### PR TITLE
feat: Add the function to hide header for key-auth plugin

### PR DESCRIPTION
**Pull request by bin-ya**
### Description

Add the function of hiding header for key-auth plugin

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by spacewander** [comment]
Let's merge master to make CI pass.

File: apisix/plugins/key-auth.lua [hunk 1]
```diff
@@ -110,6 +114,10 @@ function _M.rewrite(conf, ctx)
     end
     core.log.info("consumer: ", core.json.delay_encode(consumer))
 
+    if conf.hide_credentials then
+        core.request.set_header(ctx, "apikey", nil)
```
Comment (tzssangglass) [inline apisix/plugins/key-auth.lua:118]:
```suggestion
        core.request.set_header(ctx, conf.header, nil)
```

**Comment by tzssangglass** [comment]
hi @bin-ya PTAL https://github.com/apache/apisix/runs/5624966335?check_suite_focus=true

**Review by tzssangglass — APPROVED** [review]

File: docs/en/latest/plugins/key-auth.md [hunk 2]
```diff
@@ -41,6 +41,7 @@ For route side:
 | ---- | ------ | ----------- | ------- | ----- | ---------------------------------------------------------------------------- |
 | header  | string | optional    | apikey        |       | the header we get the key from |
 | query   | string | optional    | apikey        |       | the query string we get the key from, which priority is lower than `header` |
+| hide_credentials   | bool | optional    | false        |       | Whether to pass the apikey request headers to the upstream. |
```
Comment (tokers) [inline docs/en/latest/plugins/key-auth.md:44]: It's confusing, the header is not necessary to be "apikey", it's specified by the `header`. Also, the query should also be removed.

Comment (bin-ya) [reply docs/en/latest/plugins/key-auth.md:44]: ok

Comment (bin-ya) [reply docs/en/latest/plugins/key-auth.md:44]: Query is the content modified by others. Can I delete it?

**Review by juzhiyuan — COMMENTED** [review]
Great! Would you like to share your Apache APISIX Development experience?

**Comment by tzssangglass** [comment]
oo, another file conflict @bin-ya

File: docs/zh/latest/plugins/key-auth.md [hunk 3]
```diff
@@ -41,6 +41,7 @@ router 端配置：
 | ---- | ------ | ------ | ------ | ------ | ------------------------------------------------------------------------------------------------------------- |
 | header  | string | 可选| apikey |        | 设置我们从哪个 header 获取 key。 |
 | query  | string | 可选 | apikey |        | 设置我们从哪个 query string 获取 key，优先级低于 `header` |
+| hide_credentials  | bool | 可选 | false |        | 是否将请求头传递给 upstream。 |
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/key-auth.md:44]:
```suggestion
| hide_credentials  | bool | 可选 | false |        | 是否将含有认证信息的请求头传递给 upstream。 |
```

**Comment by tzssangglass** [comment]
File conflicts have not been resolved @bin-ya

File: docs/zh/latest/plugins/key-auth.md (hunk 3 above)
Comment (bin-ya) [reply docs/zh/latest/plugins/key-auth.md:44]: ok

File: docs/en/latest/plugins/key-auth.md (hunk 2 above)
Comment (tokers) [reply docs/en/latest/plugins/key-auth.md]: You should only handle the query that used to carry the credential.

File: docs/en/latest/plugins/key-auth.md [hunk 4]
```diff
@@ -41,6 +41,7 @@ For route side:
 | ---- | ------ | ----------- | ------- | ----- | ---------------------------------------------------------------------------- |
 | header  | string | optional    | apikey        |       | the header we get the key from |
 | query   | string | optional    | apikey        |       | the query string we get the key from, which priority is lower than `header` |
+| hide_credentials   | bool | optional    | false        |       | Whether to pass the request header containing authentication information to upstream. |
```
Comment (starsz) [inline docs/en/latest/plugins/key-auth.md:44]: Need to align the `|`.

**Review by starsz — APPROVED** [review]

File: apisix/plugins/key-auth.lua [hunk 5]
```diff
@@ -110,6 +114,10 @@ function _M.rewrite(conf, ctx)
     end
     core.log.info("consumer: ", core.json.delay_encode(consumer))
 
+    if conf.hide_credentials then
```
Comment (spacewander) [inline apisix/plugins/key-auth.lua:117]: We should remove the key according to where the key comes from.

Comment (bin-ya) [reply apisix/plugins/key-auth.lua:117]: ok, I will.

File: apisix/plugins/key-auth.lua [hunk 6]
```diff
@@ -110,6 +114,12 @@ function _M.rewrite(conf, ctx)
...
+        core.request.set_header(ctx, conf.header, nil)
+        core.request.get_uri_args(ctx)
+        core.request.set_uri_args(ctx, {})
```
Comment (tzssangglass) [inline apisix/plugins/key-auth.lua:120]: What is the purpose of doing this?

Comment (bin-ya) [reply apisix/plugins/key-auth.lua:120]: Hide information from query

Comment (tzssangglass) [reply apisix/plugins/key-auth.lua:120]: I think we don't necessary to remove the all query params
If the key is in the header, we only need to delete the key in the headers.
If the key is in the query, we only need to delete the key in the query params.

cc @spacewander right?

Comment (spacewander) [reply apisix/plugins/key-auth.lua:120]: Yes, we should only remove the key.

Comment (bin-ya) [reply apisix/plugins/key-auth.lua:120]: OK, I'll revise it right away.

**Review by tzssangglass — APPROVED** [review]

**Review by starsz — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**